src/hardware/pcspeaker.cpp

```diff
@@ -21,16 +21,43 @@
 #include <algorithm>
 #include <array>
 #include <cmath>
+#include <limits>
 
 #include "timer.h"
 #include "setup.h"
 #include "pic.h"
 
 #define SPKR_ENTRIES 1024
-#define SPKR_POSITIVE_VOLTAGE 5000.0f
-#define SPKR_NEUTRAL_VOLTAGE  0.0f
-#define SPKR_NEGATIVE_VOLTAGE -SPKR_POSITIVE_VOLTAGE
-#define SPKR_SPEED (SPKR_POSITIVE_VOLTAGE * 2.0f / 0.070f)
+#define SPKR_SPEED   (Amplitude::positive * 2.0f / 0.070f)
+
+/*
+Amplitude levels for the speaker are computed as follows:
+1. Start with the maximum allow level
+2. Reduce it by -5 LUFS to match generally upper-bound normalization
+   levels for normal audio. Knowing that the PC speaker hits peak
+   amplitude, we simply take off 5 dB.
+2. Given the PC speaker's PWM-mode doesn't actually produce sine
+   curves and instead are stepped approximations, we drop this by the
+   RMS level of a sine curve, which should provide roughly the same
+   perceived loudness.
+3. When the PC Speaker generates square waves, we apply the above reductions
+   again plus an additional RMS factor, knowing that square wave are
+   exclusively at full amplitude, and carry twice the power of sine waves.
+
+TODO: sqrt() and log() are not available as a constexpr in C++14,
+      so hard-code them for now.
+*/
+struct Amplitude {
+	static constexpr float minus_5db = 0.562341f;
+	static constexpr float sine_rms_v = 1.0f / 1.41421356237f; // 1/2^0.5
+	static constexpr float positive = std::numeric_limits<int16_t>::max() *
+	                                  minus_5db * sine_rms_v;
+	static constexpr float negative = std::numeric_limits<int16_t>::min() *
+	                                  minus_5db * sine_rms_v;
+	static constexpr float neutral = (positive + negative) / 2;
+	static constexpr float square_wave_reducer = minus_5db * sine_rms_v *
+	                                             sine_rms_v;
+};
```

Comment:
Another option would be to manually compute the values offline and describe it all in the comments, and then simply assign the actual magic numbers in the code instead of computing them at-compile-time w/ constexpr.

I opted for the constexpr because it's easy to adjust, and it works from first principles. Ideally RMS and DBs would be computed too (and I tried) but getting `log()` and `sqrt()` as constexpr's in C++14 takes way more effort than it's worth (atleast for this single case..)

Reply:
Honestly, I don't like that we're adding a new struct with all static fields - I don't think it helps with readability.

I think writing it such way would be better:
```C++
constexpr float AMPLITUDE_MINUS_5DB = 0.562341;
```
About `sine_rms_v` - what is this about? Why `sine` in the name of this constant?
If it's supposed to be the number as calculation and comment indicates, then we can use `M_SQRT1_2` now :)

Reply:
> About `sine_rms_v` Why sine in the name of this constant?

When we think of audio RMS, I immediately think of ~0.707 - but RMS is root of the mean's squared _of actual audio samples_.   If we feed the RMS calculation a square wave, it will be 1.0 because it goes from 100% to -100% "instantly" and spends no time at lower amplitudes.  So number we know and love, 0.707 is when RMS is fed perfect sine waves; which is really what most natural sounds are composed of.

So in this case, we want to scale out square-waves down to the RMS amplitudes as if they were nice sine waves :-) 

All that said; I agree it's still confusing to leave 'sine' in there, given everyone expect 0.707 anyway :-) I just wanted to clarify I'm using RMS of sine waves.